clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -128,14 +173,81 @@ public void done(long baseOffset, long logAppendTime, RuntimeException exception
         produceFuture.done();
     }
 
+    public Deque<ProducerBatch> split(int batchSize) {
+        Deque<ProducerBatch> batches = new ArrayDeque<>();
+        MemoryRecords memoryRecords = recordsBuilder.build();
+        Iterator<MutableRecordBatch> recordBatchIter = memoryRecords.batches().iterator();
+        if (!recordBatchIter.hasNext())
+            throw new IllegalStateException("Cannot split an empty producer batch.");
+        RecordBatch recordBatch = recordBatchIter.next();
+        if (recordBatchIter.hasNext())
+            throw new IllegalStateException("A producer batch should only have one record batch.");
+
+        Iterator<Thunk> thunkIter = thunks.iterator();
+        // We always allocate batch size because we are already splitting a big batch.
+        // And we also Retain the create time of the original batch.
+        ProducerBatch batch = null;
+        for (Record record : recordBatch) {
+            assert thunkIter.hasNext();
+            Thunk thunk = thunkIter.next();
+            if (batch == null) {
+                batch = createBatchOffTheBookForRecord(this.topicPartition, this.recordsBuilder.compressionType(),
+                                                       record, batchSize, this.createdMs);
+            }
+
+            // A newly created batch can always host the first message.
+            if (!batch.tryAppend(record.timestamp(), record.key(), record.value(), record.headers(), thunk)) {
+                batch.close();
+                batches.add(batch);
+                batch = createBatchOffTheBookForRecord(this.topicPartition, this.recordsBuilder.compressionType(),
+                                                       record, batchSize, this.createdMs);
+                batch.tryAppend(record.timestamp(), record.key(), record.value(), record.headers(), thunk);
+            }
+        }
+        // Close the last batch and add it to the batch list after split.
+        if (batch != null) {
+            batch.close();
+            batches.add(batch);
+        }
+        produceFuture.set(ProduceResponse.INVALID_OFFSET, NO_TIMESTAMP, new RecordBatchTooLargeException());
+        produceFuture.done();
+        return batches;
+    }
+
+    private ProducerBatch createBatchOffTheBookForRecord(TopicPartition tp,
```

Comment:
Because the split happens rarely, it seems that this sensor is going to be 0 most of the times. If there are frequent split, we will see that from the split rate sensor. So I am not sure if the sensor is needed. Also this sensor is not in the original KIP. Maybe we can add it later if it turns out to be useful.